### PR TITLE
dialog_window_refactoring code review

### PR DESCRIPTION
**Pull request by chrilehner**
related issues:   

### Pre merge/review checklist
* [ ] branch is up-to-date with the branch to be merged with, i.e. master
* [ ] travis build works
* [ ] internal code review by requester is done:
  * [ ] code is formatted
  * [ ] removed debug output `console.log`, ...
  * [ ] use of `readonly` for readonly fields and interfaces fields
  * [ ] parameters, fields, variables that are not initialized are typed, e.g.,  `function f(a)` => `function f(a: number)` 
  * [ ] optional options objects are typed using option interfaces, see e.g. `ILineUpConfig`
  * [ ] code is formatted using your preferred editor formatting tool (e.g. atom beautifier, WebStorm Format Code,...)
  * [ ] use of meaningful names in camelCase especially for public functions and fields
  * [ ] file structure: avoid of large code files containing multiple elements (classes, functions) better to create a nested structure, e.g., `model`
  * [ ] in case of DOM elements: assigning the `id` is avoided or the prefix in the lineup config is prepended
  * [ ] in case of css styles: global classes prepended with `lu-` or within a `lu-` namespace
  * [ ] public fields/function those purpose is not obvious are documented
  * [ ] complex code is documented
  * [ ] code is well structured regarding use of variable names, complexity
  * [ ] public interfaces (e.g. public interface of a class) are minimal, no superfluous fields, functions
  * [ ] complex code is not duplicated. Otherwise justified within the code, why needed
 * [ ] code reviewer is assigned
 * [ ] in case of second review round: review comments are justified or implemented

### CODE REVIEW

File: src/ui_dialogs.ts [hunk 1]
```diff
@@ -21,6 +21,71 @@ import DataProvider from './provider/ADataProvider';
 import {IBoxPlotColumn, SORT_METHOD} from './model/BoxPlotColumn';
 
 
+abstract class ADialog {
```
Comment (sgratzl) [inline src/ui_dialogs.ts]: since you already extracting things -> please extract this to an own file, too

File: src/ui_dialogs.ts [hunk 2]
```diff
@@ -21,6 +21,71 @@ import DataProvider from './provider/ADataProvider';
...
+
+  private readonly column: Column;
+  private readonly attachment;
+  private readonly title: string;
+
+
+  constructor(column, $header, title) {
```
Comment (sgratzl) [inline src/ui_dialogs.ts]: I personally prefer the following style: 
 ```Typescript
constructor(private readonly column: Column, private readonly attachment: d3.Selection<Column>, private readonly title: string)
```

File: src/ui_dialogs.ts [hunk 3]
```diff
@@ -21,6 +21,71 @@ import DataProvider from './provider/ADataProvider';
 import {IBoxPlotColumn, SORT_METHOD} from './model/BoxPlotColumn';
 
 
+abstract class ADialog {
+
+  private readonly column: Column;
```
Comment (sgratzl) [inline src/ui_dialogs.ts]: since this class doesn't use the `column` why storing it in this class anyhow?

File: src/dialogs/EditLinkDialog.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        ${this.templates.length > 0 ? 'list="ui' + this.idPrefix + 'lineupPatternList"' : ''}
+      ><br>`;
+    if (this.templates.length > 0) {
+      t += '<datalist id="ui${idPrefix}lineupPatternList">' + this.templates.map((t) => `<option value="${t}">`) + '</datalist>';
+    }
+
+    const popup = this.makePopup(t);
+
+    const that = this;
+    popup.select('.ok').on('click', function () {
```
Comment (sgratzl) [inline src/dialogs/EditLinkDialog.ts]: arrow function could be used, to avoid `that`

File: src/dialogs/SearchDialog.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import Column from '../model/Column';
+import ADialog from '../ui_dialogs';
+import DataProvider from '../provider/ADataProvider';
+
+
+
+export default class SearchDialog extends ADialog {
+  private readonly provider;
```
Comment (sgratzl) [inline src/dialogs/SearchDialog.ts]: avoid fields without types otherwise the compiler can't check those fields.

Comment (sgratzl) [inline src/dialogs/SearchDialog.ts]: same goes for function arguments

File: src/dialogs/SearchDialog.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      if (search.length >= 3) {
+        const isRegex = popup.select('input[type="checkbox"]').property('checked');
+        if (isRegex) {
+          search = new RegExp(search);
+        }
+        that.provider.searchAndJump(search, that.getColumn());
+      }
+    });
+
+    function updateImpl() {
```
Comment (sgratzl) [inline src/dialogs/SearchDialog.ts]: you could write as an alternative `const updateImple = () => ...`. Then you are sure about the this context

File: src/dialogs/EditLinkDialog.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  private readonly templates;
+  private readonly idPrefix;
+
+  constructor(column: LinkColumn, $header: d3.Selection<Column>, templates: string[] = [], idPrefix: string, title: string = 'Edit Link ($ as Placeholder)') {
+    super(column, $header, title);
+
+    this.templates = templates;
+    this.idPrefix = idPrefix;
+
+    this.openDialog();
```
Comment (sgratzl) [inline src/dialogs/EditLinkDialog.ts]: I would avoid opening the dialog directly within the constructor. At least that is something I wouldn't expect, if I'm just creating the dialog.

**Comment by chrilehner** [comment]
@sgratzl could you give me your feedback again?

File: src/dialogs/ADialog.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+import {Selection, select, event as d3event, behavior} from 'd3';
+
+
+abstract class ADialog {
+
+  constructor(private readonly _attachment: Selection<any>, private readonly title: string) {}
+
+  abstract openDialog();
+
+  get attachment() {
```
Comment (sgratzl) [inline src/dialogs/ADialog.ts]: since it is just declared `readonly` anyhow the field can be made public

File: src/dialogs/ScriptEditDialog.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  constructor(private readonly column: ScriptColumn, $header: d3.Selection<Column>, title: string = 'Edit Script') {
+    super($header, title);
+  }
+
+  openDialog() {
+    const bak = this.column.getScript();
+    const $popup = this.makePopup(`Parameters: <code>values: number[], children: Column[]</code><br>
+      <textarea autofocus="true" rows="5" autofocus="autofocus" style="width: 95%;">${this.column.getScript()}</textarea><br>`);
+
+    const updateData = (script) => {
```
Comment (sgratzl) [inline src/dialogs/ScriptEditDialog.ts]: please add typings

File: src/ui_dialogs.ts [hunk 10]
```diff
@@ -78,435 +64,12 @@ export function makePopup(attachement: Selection<any>, title: string, body: stri
...
-}
-
-
 /**
  * opens a dialog for editing the script code
  * @param column the column to edit
  * @param $header the visual header element of this column
  */
+// TODO: A refactored version of this function is in src/dialogs/ScriptEditDialog.ts. Delete the following function after the refactored version has been tested.
 export function openEditScriptDialog(column: ScriptColumn, $header: d3.Selection<Column>) {
```
Comment (sgratzl) [inline src/ui_dialogs.ts]: please delete

File: src/ui_dialogs.ts [hunk 11]
```diff
@@ -5,20 +5,10 @@
  */
```
Comment (sgratzl) [inline src/ui_dialogs.ts]: this whole file can be deleted, can't it?

**Pull request merged**